File: hashtables/hashmaps/C/src/hashmap.c
#include "hashmap.h"
#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>

HashMap *hashmap_create(int initial_size, double max_load_factor) {
  if (initial_size <= 0 || (initial_size & (initial_size - 1)) != 0) {
    return NULL;
  };

  HashMap *new_map = malloc(sizeof(HashMap));

  if (new_map == NULL) {
    return NULL;
  };

  Node **map = calloc(initial_size, sizeof(Node *));

  if (map == NULL) {
    return NULL;
  };

  new_map->map = map;
  new_map->num_elements = 0;
  new_map->size = initial_size;
  new_map->max_load_factor = max_load_factor;

  return new_map;
};

void hashmap_destroy(HashMap *map) {
  if (map == NULL) {
    return;
  };

  if (map->map == NULL) {
    free(map);
    return;
  };

  Node *head;
  Node *next;
  for (int i = 0; i < map->size; i++) {
    head = map->map[i];
    while (head != NULL) {
      next = head->next;
      free(head);
      head = next;
    };
  };

  return;
};

bool hashmap_add(HashMap *map, int key, int value) {
  if (map == NULL) {
    return false;
  };

  int p = __builtin_ctz(map->size);
  int index = knuth_hash(key) >> (32 - p);
  Node *head = map->map[index];
  Node *new_node;
  if (head == NULL) {
    new_node = malloc(sizeof(Node));
    new_node->data[0] = key;
    new_node->data[1] = value;
    new_node->next = NULL;
    map->map[index] = new_node;
    map->num_elements++;
    return true;
  };

  while (head->next != NULL) {
    if (head->data[0] == key) {
      head->data[1] = value;
      return true;
    };
    head = head->next;
  };

  if (head->data[0] == key) {
    head->data[1] = value;

  } else {
    new_node = malloc(sizeof(Node));
    new_node->data[0] = key;
    new_node->data[1] = value;
    new_node->next = NULL;
    head->next = new_node;
    map->num_elements++;
  };

  return true;
};

bool hashmap_remove(HashMap *map, int key) {
  if (map == NULL) {
    return false;
  };

  int p = __builtin_ctz(map->size);
  int index = knuth_hash(key) >> (32 - p);
  Node *head = map->map[index];
  Node *next;
  if (head == NULL) {
    return false;
  };

  next = head->next;

  if (head->data[0] == key) {
    free(head);
    map->map[index] = next;
    map->num_elements--;
    return true;
  };

  while (next != NULL) {
    if (next->data[0] == key) {
      head->next = next->next;
      free(next);
      map->num_elements--;
      return true;
    };
    head = next;
    next = next->next;
  };

  return false;
};

File: hashtables/hashmaps/Py/hashmap.py
class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.next: Node | None = None

class HashMap:
    def __init__(self, size: int, max_load_factor: int):
        if size <= 0 or (size & (size - 1)) != 0:
            raise ValueError("size must be a power of 2")
        self.size = size
        self.max_load_factor = max_load_factor
        self.map = [None] * size
        self.num_elements = 0
    
    def __trailing_zeros(self, value: int) -> int:
        n = value
        num_zeros = 0
        while n > 1:
            n //= 2
            num_zeros += 1
        return num_zeros
    
    def __knuth_hash(self, value: int) -> int:
        return (value * 2654435761) & 0xFFFFFFFF
    
    def get(self, key: int) -> int | None:
        p = self.__trailing_zeros(self.size)
        index = self.__knuth_hash(key) >> (32 - p)
        node = self.map[index]
        if node is None:
            return None
        
        while node is not None:
            if node.key == key:
                return node.value
            node = node.next
        
        return None
    
    def __resize(self) -> None: 
        new_size = 2 * self.size
        p = self.__trailing_zeros(new_size)
        if p > 30:
            return None
    
        new_map = [None] * new_size
        head = None
        next = None
        new_head = None
        new_index = None
        for i in range(0, self.size):
            head = self.map[i]
            while head is not None:
                next = head.next
                new_node = Node(head.key, head.value)
                new_index = self.__knuth_hash(head.key) >> (32 - p)
                new_head = new_map[new_index]
                if new_head is None:
                    new_map[new_index] = new_node
                else:
                    while new_head.next is not None:
                        new_head = new_head.next
                    new_head.next = new_node
                head = next
        
        self.size = new_size
        self.map = new_map


    def add(self, key: int, value: int) -> None:

        p = self.__trailing_zeros(self.size)
        index = self.__knuth_hash(key) >> (32 - p)
        head = self.map[index]
        new_node = None
        if head is None:
            new_node = Node(key, value)
            self.map[index] = new_node
            self.num_elements+=1
            if (self.num_elements / self.size) > self.max_load_factor:
                self.__resize()
            return None
        
        while head.next is not None:
            if head.key == key:
                head.value = value
                return None
            head = head.next
        
        if head.key == key:
            head.value = value
        else:
            new_node = Node(key, value)
            head.next = new_node
            self.num_elements+=1
            if (self.num_elements / self.size) > self.max_load_factor:
                self.__resize()
        return None
    
    def remove(self, key: int) -> bool:
        p = self.__trailing_zeros(self.size)
        index = self.__knuth_hash(key) >> (32 - p)
        head = self.map[index]
        next = None
        if head is None:
            return False
        
        next = head.next

        if head.key == key:
            self.map[index] = next
            self.num_elements-=1
            return True
        
        while next is not None:
            if next.key == key:
                head.next = next.next
                self.num_elements-=1
                return True
            head = next
            next = next.next
        
        return False
    
    

            